lib/compress/zstdmt_compress.c

```diff
@@ -507,8 +509,14 @@ static int ZSTDMT_serialState_reset(serialState_t* serialState, ZSTDMT_seqPool*
         memset(serialState->ldmState.hashTable, 0, hashSize);
         memset(serialState->ldmState.bucketOffsets, 0, bucketSize);
     }
-    serialState->params = params;
+    serialState->params = *paramsPtr;
     serialState->params.jobSize = (U32)jobSize;
+
+    /* Update window state and fill hash table with dict */
+    if (paramsPtr->ldmParams.enableLdm && dict) {
+        ZSTD_window_update(&serialState->ldmState.window, dict, dictSize);
+        ZSTD_ldm_fillHashTable(&serialState->ldmState, (const BYTE*)dict, (const BYTE*)dict + dictSize, &paramsPtr->ldmParams);
+    }
```

Comment:
>I believe it's worth underlining what needs to be updated, and why it's not possible to do it differently, or why it's the better choice.

If we want to prevent prefs from being changed by this function, we'd just have to move this if statement outside of it. I placed it in here originally because it seemed like a 'reset' type operation but looking back, moving this outside and keeping the rest as before might be a better solution.

Reply:
Transforming an argument from "read only" into "read-write" is merely an early warning sign.
It's not an "automatic proof" that it's "bad", but the implications should be clearly understood.
In particular, where does it make more sense to do those changes, and what happens to those changes afterwards ?

In order to select between 2 implementation models, 
one has to understand why this function is invoked,
hence look at the places where it's invoked.

Right now, I see 2 invocations : in `ZSTDMT_initCStream_internal()`, and in ZSTDMT_compress_advanced_internal()` (which feels logical retrospectively).

Interestingly, in both cases, `params` is a read argument from the calling function,
which means the _changes enforced into `params` will not be propagated beyond the function_.

I presume it's fine, and exactly what's needed for `ZSTDMT_compress_advanced_internal()`.

However, for `ZSTDMT_initCStream_internal()`, it seems more strange : 
the function returns right after invoking `ZSTDMT_serialState_reset()`, 
which means that any change enforced into `params` is basically useless, since `params` becomes out of scope immediately after.

`ZSTDMT_initCStream_internal()` is the most important variant used by the `zstd` cli for most files (all non-tiny ones).

This suggests that _currently_, the changes made into `params.ldmParams` are basically cancelled immediately afterwards, hence not applied during the following compression session.
I would expect such an impact to be visible during tests ?

Reply:
>This suggests that currently, the changes made into params.ldmParams are basically cancelled immediately afterwards, hence not applied during the following compression session.

```
serialState->params = *paramsPtr;
```

is set later in that same function btw so that is how the changes propagate. So params becomes useless afterwards yes but it changes the internal params after its used as a temp variable or sorts it looks like.

Reply:
Going to make params immutable